### PR TITLE
Add commandline params for tasks

### LINKED ISSUE #503: Add command line args support for tasks
When work for https://github.com/spring-cloud/spring-cloud-deployer/issues/20 is done, its wingman here needs to similar things in a shell and controllers.

### LINKED ISSUE #523: Spring Cloud Task launch properties are not persisted
When launching a task, launch properties are not persisted to the database. 

```
task launch scdf-batch-simple --properties sample.jobParams={"throwError":"true","-secondParam(long)":"99999999"}
```

### PR DESCRIPTION
**Pull request by jvalkeal**
- Modify shell launch command.
- Add raw list of params throughout a chain
  from shell into a deployer SPI call.
- Added one test verifying params.
- In DefaultTaskJobService.restartJobExecution() params are
  coming from TaskExecution, theory at least, but this is not
  testable right now as we don't have restart command.
- Fixes #503
- Fixes #523

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java [hunk 1]
```diff
@@ -63,11 +65,13 @@ public TaskDeploymentController(TaskService taskService) {
...
 	 * 					 key=value pairs
+	 * @param params the runtime commandline arguments
 	 */
 	@RequestMapping(value = "/{name}", method = RequestMethod.POST)
 	@ResponseStatus(HttpStatus.CREATED)
-	public void deploy(@PathVariable("name") String taskName, @RequestParam(required = false) String properties) {
-		this.taskService.executeTask(taskName, DeploymentPropertiesUtils.parse(properties));
+	public void deploy(@PathVariable("name") String taskName, @RequestParam(required = false) String properties,
+			@RequestParam(required = false) List<String> params) {
+		this.taskService.executeTask(taskName, DeploymentPropertiesUtils.parse(properties), params);
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: probably need to parse each of the parameters (delimited by space) out into a list.   Else it will send the all params as a single param instead of collection parameters.  
```
		List<String> paramsToUse = new ArrayList<>();
		if(params != null) {
			for (String paramsItem : params) {
				for (String param : StringUtils.delimitedListToStringArray(paramsItem, " ")) {
					paramsToUse.add(param);
				}
			}
		}
		this.taskService.executeTask(taskName, DeploymentPropertiesUtils.parse(properties), paramsToUse);
```

Comment (mminella) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: Why is this needed?

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/TaskControllerTests.java [hunk 2]
```diff
@@ -222,4 +224,31 @@ public void testLaunch() throws Exception {
...
 	}
+
+	@Test
+	public void testLaunchWithParams() throws Exception {
+		repository.save(new TaskDefinition("myTask2", "foo2"));
+		this.registry.register("task.foo2", new URI("maven://org.springframework.cloud:foo2:1"));
+
+		mockMvc.perform(
+				post("/tasks/deployments/{name}", "myTask2")
+				.param("params", "--foobar=jee")
```
Comment (mminella) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/TaskControllerTests.java]: Two questions here:

* How does this look with multiple params?
* why would this go over the wire with `--`?  Is that how other options are sent (pardon my ignorance on this)?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java (hunk 1 above)
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: Good question!

When executing from the command line the following arguments `—foo=bar —baz=boo`  would have 2 elements in the arg array.
`—foo=bar`
`—baz=boo`
while a task launched from SCDF would have only 1 element in the arg array.
`—foo=bar —baz=boo`

[

Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: I thought about trying to parse args but as we really don't know what's going to get passed in I felt safer not trying split it into a list. It really is a raw string(having arbitry params) what user can pass in via shell.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/TaskControllerTests.java (hunk 2 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/TaskControllerTests.java]: As mentioned above, we don't care a format what user is passing in. With `--` that'd be a boot property but I don't know if batch itself supports any type of command line args.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java (hunk 1 above)
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: I think the primary issue is that without us parsing the args in SCDF, the behavior of how the Task will have to parse the args will be different based on how the task was launched.( i.e. from command line will be multiple elements and from SCDF only one element.)   The way a task parses the args should be consistent, regardless of how it was launched.

Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: Lets take a default `format` property and pass it in via params as `--format="yyyy-MM-dd HH:mm:ss.SSS"`.

It has a space in it between `dd` and `HH`. We can't just start to split these with spaces because then one parameter would be splitted into to two parameters.

I'm now also seeing some weird behaviour with local deployer where multiple params gets mixed up somehow. Something to do with escaping or ways how stuff is started.

**Comment by jvalkeal** [comment]
So trouble with `ProcessBuilder` in local deployer.
```
task launch --name footask --params "--format='yyyy-MM-dd HH:mm:ss.SSS'"
```
is causing broken format in logs
```
2016-05-12 13:58:31.829  INFO 31548 --- [           main] o.s.c.t.a.t.TimestampTaskConfiguration   : yyyy-MM-dd HH:mm:ss.SSS
```
Process launched as seeing from ps is
```
java -jar /home/jvalkealahti/.m2/repository/org/springframework/cloud/task/app/timestamp-task/1.0.0.M1/timestamp-task-1.0.0.M1.jar --format='yyyy-MM-dd HH:mm:ss.SSS'
```
If I run that command manually I get correct format in logs
```
2016-05-12 14:00:54.460  INFO 31723 --- [           main] o.s.c.t.a.t.TimestampTaskConfiguration   : 2016-05-12 14:00:54.460
```
If I add second argument after it breaks with error:
```
task launch --name footask --params "--format='yyyy-MM-dd HH:mm:ss.SSS' --foo=bar"
```
```
Caused by: java.lang.IllegalArgumentException: Illegal pattern character 'f'
	at java.text.SimpleDateFormat.compile(SimpleDateFormat.java:826) ~[na:1.8.0_45]
```

This may be something mentioned in SO http://stackoverflow.com/questions/12124935/processbuilder-adds-extra-quotes-to-command-line

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java (hunk 1 above)
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: I know there is a better way to do this but:
```
		Pattern regex = Pattern.compile("(\\s(?=([^\\\"']*[\\\"'][^\\\"']*[\\\"'])*[^\\\"']*$))");
		for( String paramsItem : params) {
			Matcher regexMatcher = regex.matcher(paramsItem);
			int start = 0;
			while (regexMatcher.find()) {
				paramsToUse.add(paramsItem.substring(start, regexMatcher.start()));
				start = regexMatcher.start();
			}
			if(paramsItem != null && paramsItem.length() > 0){
				paramsToUse.add(paramsItem.substring(start,paramsItem.length()));
			}
		}```
Launching this task:   task launch foo --params "--format=\"yyyy-MM-dd HH:mm:ss.SSS\" bifflediffle=snizersnot wombats=\"eat veggies\" noway=really"

displayed this in my sample task: 
 you got ==>--format="yyyy-MM-dd HH:mm:ss.SSS"
 you got ==> bifflediffle=snizersnot
 you got ==> wombats="eat veggies"
 you got ==> noway=really

**Comment by jvalkeal** [comment]
ok, so we need to find a way to split that raw command-line arguments string. Just checked a process manually:
```
$ pid=$(ps -ef|grep timestamp|grep -v grep | awk '{print $2}'); sudo strings /proc/$pid/cmdline
java
-jar
/home/jvalkealahti/.m2/repository/org/springframework/cloud/task/app/timestamp-task/1.0.0.M1/timestamp-task-1.0.0.M1.jar
--format='yyyy-MM-dd HH:mm:ss.SSS' --foo=bar
```
In above `--format` would get value `'yyyy-MM-dd HH:mm:ss.SSS' --foo=bar` and things go south.
If I run it from command line its
```
$ pid=$(ps -ef|grep timestamp|grep -v grep | awk '{print $2}'); sudo strings /proc/$pid/cmdline
java
-jar
/home/jvalkealahti/.m2/repository/org/springframework/cloud/task/app/timestamp-task/1.0.0.M1/timestamp-task-1.0.0.M1.jar
--format=yyyy-MM-dd HH:mm:ss.SSS
--foo=bar
```

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java (hunk 1 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: Do we need to support arbitrary args which are not in `key=value` format?

**Comment by cppwfs** [comment]
@jvalkeal I did a similar test where I created a timestamp task that printed the format.   And there are 2 problems.  

[Successful Run with multiple parameters from command line]
* Running from the command line with the following `java -jar timestamp-task/target/timestamp-task-1.0.0.BUILD-SNAPSHOT.jar --format="yyyy-MM-dd HH:mm:ss.SSS" --foo=bar`
* Running from the command line with the following `java -jar timestamp-task/target/timestamp-task-1.0.0.BUILD-SNAPSHOT.jar --format='yyyy-MM-dd HH:mm:ss.SSS' --foo=bar`
The task completes with the expected results.  i.e. 2 parameters work

[Running from SCDF]
Problem 1: Using existing code it is placing both --format &  --foo parameters into the format property.  Using the code sample I enumerated above fixes that issue.  
Problem 2: Like you said when executing from SCDF the " or ' are left in the format field, thus causing the logging problem.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java (hunk 1 above)
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: I'm going to say yes.  Because a `task` is kind of the wild west it can be any boot app.  But I defer to @mminella  on this one.

**Comment by jvalkeal** [comment]
Ok, I'll tweak that stuff based on what you did. That kinda locks in a format we can use so need to document it. I'll do that tomorrow.

**Comment by jvalkeal** [comment]
I think we need to take this back into a drawing board. From dataflow shell if we need to have a space, we either need to use `--format=\"yyyy-MM-dd HH:mm:ss.SSS\"`, `--format='yyyy-MM-dd HH:mm:ss.SSS', or `'--format=yyyy-MM-dd HH:mm:ss.SS'`.

With all cases we've left double/single quotes which are then passed to ProcessBuilder which then puts those into a real args and things fail. If those are used in a real OS shell, system takes those away.

This also makes it difficult to really know how underlying deployer(local, pcf, yarn) should really handle those command line args. With local those goes directly into ProcessBuilder, with yarn it's really a shell command to be executed.

This raw command-line args is not going to fly, we need to find something more generic and reliable.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java (hunk 1 above)
Comment (mminella) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskDeploymentController.java]: I'd say we need to support anything that boot supports.  I may be wrong here, but I don't think they support flag args (`--foo`), but only properties (`--foo=bar`).  This is something we'd need to confirm with @philwebb

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/TaskControllerTests.java (hunk 2 above)
Comment (mminella) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/TaskControllerTests.java]: When running a batch app via boot, all the args are passed in as job parameters.

**Comment by jvalkeal** [comment]
wohoo, closing this one and planning for 3rd PR.
